### PR TITLE
PR for discussion of implementing ordersubmission without nonce

### PR DESCRIPTION
**Pull request by josojo**

### CODE REVIEW

File: contracts/SnappAuction.sol [hunk 1]
```diff
@@ -11,17 +11,11 @@ contract SnappAuction is SnappBase {
...
     }
 
-    struct StandingOrderData {
-        mapping(uint => StandingOrderBatch) reservedAccountOrders;
-        uint standingOrderNonce;
-    }
-
-    //mapping from accountId to nonce to StandingOrderBatch
-    mapping (uint16 => StandingOrderData) public standingOrderPlacements;
+    //mapping from accountId to validFromAuctionIndex to StandingOrderBatch
```
Comment (bh2smith) [inline contracts/SnappAuction.sol]: can't this just be standing orders? I don't see the reason for the word "Placement" here

File: contracts/SnappAuction.sol [hunk 2]
```diff
@@ -11,17 +11,11 @@ contract SnappAuction is SnappBase {
...
+    mapping (uint16 => mapping(uint => StandingOrderBatch)) public standingOrderPlacements;
```
Comment (bh2smith) [inline contracts/SnappAuction.sol:18]: I don't see the reason for the word "Placements" here. Can we please just have "standingOrders"

**Review by bh2smith — COMMENTED** [review]
I really don't like all these long words: Here are my suggestions in order of priority.

- prevValidFromAuctionIndex --> prevAuctionFrom
- standingOrderPlacements --> standingOrders
- validFromAuctionIndex --> auctionFrom
- validToAuctionIndex --> auctionTo

**Comment by bh2smith** [comment]
Also, if we are going to start using `userId` I would think that every use of `accountId` should also be changed everywhere. That is, we need to use one or the other consistently, but not a mix of both. No strong preference either way.

File: contracts/SnappAuction.sol [hunk 3]
```diff
@@ -79,21 +73,14 @@ contract SnappAuction is SnappBase {
         return auctions[slot].appliedAccountStateIndex != 0;
     }
 
-    function getStandingOrderHash(uint16 userId, uint128 nonce) public view returns (bytes32) {
-        return standingOrderPlacements[userId].reservedAccountOrders[nonce].orderHash;
+    function getStandingOrderHash(uint16 userId, uint validFromAuctionIndex) public view returns (bytes32) {
```
Comment (bh2smith) [inline contracts/SnappAuction.sol:76]: I don't see that this function is actually invoked anywhere. Is it needed?

File: contracts/SnappAuction.sol [hunk 4]
```diff
@@ -79,21 +73,14 @@ contract SnappAuction is SnappBase {
...
+        return standingOrderPlacements[userId][validFromAuctionIndex].orderHash;
     }
     
-    function getStandingOrderValidFrom(uint16 userId, uint128 nonce) public view returns (uint) {
-        return standingOrderPlacements[userId].reservedAccountOrders[nonce].validFromAuctionIndex;
+    function getStandingOrdervalidToAuctionIndex(uint16 userId, uint validFromAuctionIndex) public view returns (uint) {
```
Comment (bh2smith) [inline contracts/SnappAuction.sol:80]: Same. Is this required for anything? Do we plan to use it for unit testing?

File: contracts/SnappAuction.sol [hunk 5]
```diff
@@ -102,13 +89,17 @@ contract SnappAuction is SnappBase {
...
+        uint128[] memory sellAmounts,
+        uint prevValidFromAuctionIndex
     ) public onlyRegistered() {
         
         // Update Auction Hash based on request
         uint16 accountId = publicKeyToAccountMap(msg.sender);
         require(accountId <= AUCTION_RESERVED_ACCOUNTS, "Accout is not a reserved account");
 
+        // Check correctness of prevValidFromAuctionIndex
+        require( standingOrderPlacements[accountId][prevValidFromAuctionIndex].validToAuctionIndex != MAX_UINT || prevValidFromAuctionIndex == 0, "validToAuctionIndex is set, this is not the previous orderbatch");
```
Comment (bh2smith) [inline contracts/SnappAuction.sol:102]:
```suggestion
        require(standingOrderPlacements[accountId][prevValidFromAuctionIndex].validToAuctionIndex != MAX_UINT || prevValidFromAuctionIndex == 0, "validToAuctionIndex is set, this is not the previous orderbatch");
```

File: contracts/SnappAuction.sol [hunk 6]
```diff
@@ -127,17 +118,13 @@ contract SnappAuction is SnappBase {
...
-            standingOrderPlacements[accountId].reservedAccountOrders[currentNonce].validToAuctionIndex = auctionIndex - 1;
-            newNonce += 1;
-            standingOrderPlacements[accountId].standingOrderNonce = newNonce;
-            standingOrderPlacements[accountId].reservedAccountOrders[newNonce].validFromAuctionIndex = auctionIndex;
+        if (auctionIndex > prevValidFromAuctionIndex) {
+            standingOrderPlacements[accountId][prevValidFromAuctionIndex].validToAuctionIndex = auctionIndex - 1;
+            standingOrderPlacements[accountId][auctionIndex].validToAuctionIndex = MAX_UINT;
         }
-        standingOrderPlacements[accountId].reservedAccountOrders[newNonce].orderHash = orderHash;
+        standingOrderPlacements[accountId][auctionIndex].orderHash = orderHash;
```
Comment (bh2smith) [inline contracts/SnappAuction.sol:132]: This looks like a weird/inconsistent change that does not reflect identical logic to the previous code (which used an apparently different value).

Comment (josojo) [reply contracts/SnappAuction.sol:132]: In my head this makes sense :) 
the mapping was changed to 
accoutID => validFromAuctionIndex => StandingOrderBatch

And the new validity will be set here to (validityFromAuctionIndex= auctionId)

This all goes a little bit further than felix suggestion, as it basically removes the nonce and stores the validFrom only indirectly.

Comment (bh2smith) [reply contracts/SnappAuction.sol:132]: Cool, just making sure!

**Pull request closed without merging**